### DISCUSSION TITLE
Capabilities of the datalog reasoner?

### DISCUSSION THREAD

**Discussion by jubax** [EvgSkv/logica#31] [Q&A]
Hi, nice new language! I'm always happy when new projects use Datalog :-)

I could not find a description of the capabilities of the reasoning engine:

- Is stratified negation supported?
- Do you execute bottom-up or top-down?
- Are there optimizers like rule unfolding or magic sets?

Thank you!

**Comment by EvgSkv** [comment]
These are great questions, I'll post response soon.

**Comment by EvgSkv** [comment]
Since the language is written as a compilation from Datalog-like syntax to SQL I didn't have a chance to dive into the classical art of Datalog execution, so I needed to read a little to even give a rough answer to your questions.

* Is stratified negation supported?

Please correct me if I'm wrong, but as far as I understood, the notion of stratified negation is introduced for compilation of recursive programs and recursion is not supported in Logica yet. Some problems that are usually solved via recursion [can be solved](https://github.com/EvgSkv/logica/blob/main/examples/scripts/reachability_use.l) with tricks in Logica. Support of recursion is intended in the future.

* Do you execute bottom-up or top-down?
I believe that compilation to SQL with further SQL execution corresponds to bottom-up.

* Are there optimizers like rule unfolding or magic sets?
No, there are no such optimizations, but they could be considered in the future.

**Reply by jubax** [reply]
Thank you :-)

I was mostly asking out of professional interest as my company develops (as a part of a validation component) an in-house Datalog engine which can be used either standalone in RAM or backed by a SQL database. (We already support stratified negation, top-down execution etc.).

**Reply by EvgSkv** [reply]
Do you have public documentation for it? I'd be interested to learn about it.

**Reply by jubax** [reply]
Our reasoner was based on the IRIS reasoner from DERI Innsbruck, but over the years the whole code was basically re-written. In particular we added scalable DB integration, zero-allocation in-memory reasoning and a top-down engine.

The original website (iris-reasoner.org) seems to be down, but on this GitHub page a clone seems to exist:

https://github.com/NICTA/iris-reasoner

**Reply by EvgSkv** [reply]
Thanks for the link! Looks very powerful.

I wonder how much of the reasoning power could be expressed as transformation of the propositional formula to a logically equivalent, but computationally simpler form. It is would be easy to enable custom Logica plug-ins for this kind of transforms.

**Reply by kamahen** [reply]
You might be interested in SYLLOG by Adrian Walker. It's described in _Knowledge systems and prolog: A logical approach to expert systems and natural language processing_ byAdrian Walker (Ed.), Michael McCord, John F. Sowa and Walter G. Wilson. It appears that the Syllog section is available as a [separate reprint](https://www.amazon.com/Syllog-Knowledge-Management-Classic-Reprint/dp/1333378645). See also https://www.complang.tuwien.ac.at/ulrich/syllog.html

Syllog handles stratification and negation, and also features a simplified English syntax. There are a few articles giving examples of its use.

**Reply by EvgSkv** [reply]
Thanks Peter! Looks interesting, I'll read more about SYLLOG.

**Reply by jubax** [reply]
> Our reasoner was based on the IRIS reasoner from DERI Innsbruck, but over the years the whole code was basically re-written. In particular we added scalable DB integration, zero-allocation in-memory reasoning and a top-down engine.
> 
> The original website (iris-reasoner.org) seems to be down, but on this GitHub page a clone seems to exist:
> 
> https://github.com/NICTA/iris-reasoner

I found the new website: http://iris-reasoner.sti2.at/

Has some additional info.

**Reply by EvgSkv** [reply]
This is great! Thanks, @jubax.
